### PR TITLE
SimplifyLibCalls: Emit vector ldexp intrinsics in exp2->ldexp combine

### PR DESCRIPTION
**Pull request by arsenm**

### CODE REVIEW

File: llvm/lib/Transforms/Utils/SimplifyLibCalls.cpp [hunk 1]
```diff
@@ -2000,11 +2000,16 @@ static Value *getIntToFPVal(Value *I2F, IRBuilderBase &B, unsigned DstWidth) {
...
-    if (BitWidth < DstWidth ||
-        (BitWidth == DstWidth && isa<SIToFPInst>(I2F)))
-      return isa<SIToFPInst>(I2F) ? B.CreateSExt(Op, B.getIntNTy(DstWidth))
-                                  : B.CreateZExt(Op, B.getIntNTy(DstWidth));
+    unsigned BitWidth =
+        Op->getType()->getScalarType()->getPrimitiveSizeInBits();
+    if (BitWidth < DstWidth || (BitWidth == DstWidth && isa<SIToFPInst>(I2F))) {
+      Type *IntTy = B.getIntNTy(DstWidth);
+      if (VectorType *VT = dyn_cast<VectorType>(I2F->getType()))
+        IntTy = VectorType::get(IntTy, VT->getElementCount());
```
Comment (nikic) [inline llvm/lib/Transforms/Utils/SimplifyLibCalls.cpp:2008]: `Type::getWithNewBitWidth()`?

File: llvm/lib/Transforms/Utils/SimplifyLibCalls.cpp [hunk 2]
```diff
@@ -2000,11 +2000,13 @@ static Value *getIntToFPVal(Value *I2F, IRBuilderBase &B, unsigned DstWidth) {
     Value *Op = cast<Instruction>(I2F)->getOperand(0);
     // Make sure that the exponent fits inside an "int" of size DstWidth,
     // thus avoiding any range issues that FP has not.
-    unsigned BitWidth = Op->getType()->getPrimitiveSizeInBits();
-    if (BitWidth < DstWidth ||
-        (BitWidth == DstWidth && isa<SIToFPInst>(I2F)))
-      return isa<SIToFPInst>(I2F) ? B.CreateSExt(Op, B.getIntNTy(DstWidth))
-                                  : B.CreateZExt(Op, B.getIntNTy(DstWidth));
+    unsigned BitWidth =
+        Op->getType()->getScalarType()->getPrimitiveSizeInBits();
```
Comment (nikic) [inline llvm/lib/Transforms/Utils/SimplifyLibCalls.cpp:2004]:
```suggestion
        Op->getType()->getScalarSizeInBits();
```

**Review by nikic — APPROVED** [review]
LGTM

File: llvm/lib/Transforms/Utils/SimplifyLibCalls.cpp [hunk 3]
```diff
@@ -2000,11 +2000,13 @@ static Value *getIntToFPVal(Value *I2F, IRBuilderBase &B, unsigned DstWidth) {
     Value *Op = cast<Instruction>(I2F)->getOperand(0);
     // Make sure that the exponent fits inside an "int" of size DstWidth,
     // thus avoiding any range issues that FP has not.
-    unsigned BitWidth = Op->getType()->getPrimitiveSizeInBits();
-    if (BitWidth < DstWidth ||
-        (BitWidth == DstWidth && isa<SIToFPInst>(I2F)))
-      return isa<SIToFPInst>(I2F) ? B.CreateSExt(Op, B.getIntNTy(DstWidth))
-                                  : B.CreateZExt(Op, B.getIntNTy(DstWidth));
+    unsigned BitWidth =
+        Op->getType()->getScalarSizeInBits();
```
Comment (arsenm) [inline llvm/lib/Transforms/Utils/SimplifyLibCalls.cpp:2004]:
```suggestion
    unsigned BitWidth = Op->getType()->getScalarSizeInBits();
```

**Pull request merged**